File: code/sorting/src/topological_sort/topological_sort.swift
var inDegree: [Int] = []

func topologicalSort(adjList: [[Int]]) -> [Int] {
    var result: [Int] = []
    let numVertices = adjList.count - 1
    var queue = [Int]()
    for i in 1 ... numVertices {
        if inDegree[i] == 0 {
            queue.append(i)
        }
    }

    while !queue.isEmpty {
        let currVertex = queue.first!
        queue.removeFirst()

        result.append(currVertex)

        for i in 0 ..< adjList[currVertex].count {
            let neighbour = adjList[currVertex][i]
            inDegree[neighbour] -= 1;

            if inDegree[neighbour] == 0 {
                queue.append(neighbour)
            }
        }
    }

    return result
}

    /* Consider following graph
     * 1----->2----->5
     * |     ^^
     * |    / |
     * |   /  |
     * |  /   |
     * v /    |
     * 4----->3
     * One of the solution would be
     * 1 -> 4 -> 3 -> 2 -> 5
     */

let numVertices = 5
inDegree = Array(repeating: 0, count: numVertices + 1)
var adjList: [[Int]] = Array(repeating: [], count: numVertices + 1)

adjList[1].append(2)
adjList[1].append(4)
adjList[2].append(5)
adjList[3].append(2)
adjList[4].append(2)
adjList[4].append(3)

inDegree[1] = 0
inDegree[2] = 3
inDegree[3] = 1
inDegree[4] = 1
inDegree[5] = 1

let topSort = topologicalSort(adjList: adjList)

for num in topSort {
    print(num, terminator: "")
    if num != topSort.last! {
        print("->",  terminator: "")
    }
}

File: code/graph_algorithms/src/topological_sort/topological_sort.py
""" Part of Cosmos by OpenGenus Foundation """

from collections import deque

NUM_V = 100005
adj_list = [[] for i in range(0, NUM_V)]
in_degree = [0 for i in range(0, NUM_V)]


def toposort():
    result = []
    q = deque()

    for i in range(1, vertex + 1):
        # if no incoming degree, add to queue
        if in_degree[i] == 0:
            q.append(i)

    while q:
        front = q.popleft()

        # add into toposort result
        result.append(front)

        for neighbor in adj_list[front]:
            in_degree[neighbor] -= 1

            # if already no incoming degree, then ready to put in toposort list
            if in_degree[neighbor] == 0:
                q.append(neighbor)

    return result


"""
consider following graph
1----->2----->5
|     ^^
|    / |
|   /  |
|  /   |
v /    |
4----->3
one of the solution would be
1 -> 4 -> 3 -> 2 -> 5
"""

vertex = 5
adj_list[1].append(2)
adj_list[1].append(4)
adj_list[2].append(5)
adj_list[3].append(2)
adj_list[4].append(2)
adj_list[4].append(3)

in_degree[1] = 0
in_degree[2] = 3
in_degree[3] = 1
in_degree[4] = 1
in_degree[5] = 1
result = toposort()

print("toposort :", end="")
for vtx in result:
    print(" {:d}".format(vtx), end="")

print()
